### ISSUE TITLE
Feature Request: LAN Discovery

### ISSUE THREAD

**Issue by james7132** [ValveSoftware/GameNetworkingSockets#82]
While this library has shown to scale up to larger server based use cases and cross-continent P2P connections, the lack of a LAN discovery tool for quick and dirty LAN play makes it hard to scale down. I'm working on direct 1v1 games that will largely benefit from avoiding a hop to the nearest relay/authoritative server to establish a game over the same subnet.

**Comment by fletcherdvalve** [comment]
It depends on exactly what is meant by discovery.  If "discovery" means the ability to ask "is there anybody near me?" and get a list of peers, that might belong in this library, but maybe not.  On the other hand, if you want to connect to a peer, and you "discover" that they are on the same LAN, and so you route using local addresses instead of relaying, that I think is absolutely a good thing to have in this API.  That will probably part of #64.

I'll keep the first part, the discovery part, open under this bug.

I do realize that only having the second part means that there is some way to know that the other peer exists, and furthermore than you have some identifier you use to connect to them that is not already their LAN address.  (If you already had their LAN address you would just use UDP connect.)  So in practice it might not serve the needs you are thinking of, and you really do need the true LAN "discovery" bit, too.

**Comment by james7132** [comment]
The goal would be to find other players on the same subnet which can then
be used for direct, relay-less connection. Usually done by having one host
broadcast out to the entire subnet "Hey I have a local server running,
connect to me!" and having the others listen for those broadcasts.

The expected use case is something like LAN parties and local play that may
not necessitate being on the same physical machine

I understand that matchmaking is generally out of the scope of this
library, and that most games nowadays prefer to defer to an authoritative
remote server, but these discovery protocols often operate at the same
level as this library, and this would be immensely useful for P2P games.

**Comment by fletcherdvalve** [comment]
Yeah, that's what I thought you meant.  I agree that it is not totally out of place in this library.

Several design decisions to think through about exactly what interface the API exposes, where the server sends out broadcasts, or is only listening for broadcast searches, whether the beacon is the same port as the listen socket or is conceptually a separate entity, etc.  And also what the relationship between anything we add here and existing Steamworks features.  Does it use the Steam protocols, etc, etc.  Lots of decisions to think through.

Do you know of other APIs that you think did a good job with this?

**Comment by rjrivera** [comment]
I am late to this party; I'm trying to figure out what off-the-shelf Link-Local solutions may be integrated with the P2P subset of the GameNetworkingSockets library, and what @james7132 is describing sounds like what I want to do too, but as always the devil is in the details. I want to avoid the need for a router. 

I am starting from Link-Local and trying to walk backwards to the Steamworks GameNetworking Sockets library; does that sound appropriate for this issue? I am also looking at issue [https://github.com/ValveSoftware/GameNetworkingSockets/issues/136] when I get to the signaling service part of this. I am eyeballing how Prosody IM does the XMPP components (MIT License) but keeping the horse after the cart.
